### PR TITLE
(feat): add JIT integrator tests

### LINKED ISSUE #164: Add tests for JIT
# Feature
## Desired Behavior / Functionality
[Seems we are not testing JIT functionality](https://github.com/esa/torchquad/pull/141#issuecomment-1455757843), we should find a way to test it.

## What Needs to Be Done
- Test it :) , JIT can be used as described [here](https://torchquad.readthedocs.io/en/main/tutorial.html#compiling-the-integrate-method)

## How Can It Be Tested
🥁 with the tests

### PR DESCRIPTION
**Pull request by ilan-gold**
# Description

Adds tests for JIT integration

## Resolved Issues

- [x] fixes #164 

## How Has This Been Tested?

- [x] Added 1d JIT tests

## Related Pull Requests

- #PR

### CODE REVIEW

**Comment by ilan-gold** [comment]
@gomezzz I've added some tests for this that work.  There are a few catches.  The main thing is that we don't test repeated integration for the same integrand-N-dim JIT combination (which is the point of JIT, but maybe not something we actually need to test).  This is because the test suite doesn't really track `integrand_shape` in the right places besides whether or not the integrand is multi-dimensional in general (from #160).  So we have a few options here:

1.  Refactor the test suite to allow for finer grained control over what is run and how (i.e., more explicitly tracking `integrand_shape`) - so a different PR in addition to this one.
2. Add another step to this PR to allow for re-running the 1d tests with a single JIT call (the functionality for which already exists), as opposed to this PR where JIT happens every time.
3. Keep what we have (i.e., JIT every integrand).

**Comment by ilan-gold** [comment]
P.S One advantage of refactoring would be that we don't have to run all the tests.  I'm not sure what the goal of "unit testing" JIT is besides functionality i.e., I'm not sure we actually need to check accuracy of every single integral.  We're more interested in not-crashing.

**Comment by gomezzz** [comment]
> P.S One advantage of refactoring would be that we don't have to run all the tests. I'm not sure what the goal of "unit testing" JIT is besides functionality i.e., I'm not sure we actually need to check accuracy of every single integral. We're more interested in not-crashing.

Agree :)

> So we have a few options here:

Any of them are okay for me. At the moment the tests take ~4min which is fine and I'll take any improvement :D

File: torchquad/tests/boole_test.py [hunk 1]
```diff
@@ -61,6 +61,65 @@ def _run_boole_tests(backend, _precision):
...
     # assert error < 5e-9
 
+    # JIT Tests
+    if backend != "numpy":
+        N = 401
+        jit_integrate = None
+
+        def integrate(*args, **kwargs):
+            nonlocal jit_integrate
+            if jit_integrate is None:
```
Comment (gomezzz) [inline torchquad/tests/boole_test.py:71]: Could you add a small comment explaining the though process here? It is a bit confusing :thinking:

File: torchquad/tests/boole_test.py [hunk 2]
```diff
@@ -61,6 +61,65 @@ def _run_boole_tests(backend, _precision):
...
+            None  # set to None again so can be re-used with new integrand shape
+        )
+
+        def integrate(*args, **kwargs):
+            nonlocal jit_integrate
+            if jit_integrate is None:
+                jit_integrate = bl.get_jit_compiled_integrate(
+                    dim=1, N=N, backend=backend
+                )
+            return jit_integrate(*args, **kwargs)
```
Comment (gomezzz) [inline torchquad/tests/boole_test.py:104]: is the redeclaration necessary? was declared just above? :thinking:

Comment (gomezzz) [reply torchquad/tests/boole_test.py:104]: Same two comments apply to some of the other files as well, I think :v:

File: torchquad/tests/simpson_test.py [hunk 3]
```diff
@@ -89,6 +89,65 @@ def _run_simpson_tests(backend, _precision):
     for error in errors:
         assert error < 5e-9
 
+    if backend != "numpy":
```
Comment (gomezzz) [inline torchquad/tests/simpson_test.py:92]: comment missing that the JIT section starts here (you had one in the others)

File: torchquad/tests/trapezoid_test.py [hunk 4]
```diff
@@ -86,6 +86,65 @@ def _run_trapezoid_tests(backend, _precision):
     for error in errors:
         assert error < 7000
 
+    if backend != "numpy":
```
Comment (gomezzz) [inline torchquad/tests/trapezoid_test.py:89]: comment missing that the JIT section starts here (you had one in the others)

Comment (ilan-gold) [reply torchquad/tests/trapezoid_test.py:90]: done!

File: torchquad/tests/simpson_test.py (hunk 3 above)
Comment (ilan-gold) [reply torchquad/tests/simpson_test.py:93]: done!

File: torchquad/tests/boole_test.py (hunk 1 above)
Comment (ilan-gold) [reply torchquad/tests/boole_test.py:73]: done (i think)

**Comment by ilan-gold** [comment]
Well that now-failing test is flaky - fails sometimes, passes others.

**Comment by gomezzz** [comment]
> Well that now-failing test is flaky - fails sometimes, passes others.

let's increase the bounds on this one? 🤔

**Comment by ilan-gold** [comment]
Sounds good to me

**Review by gomezzz — APPROVED** [review]
👍

**Pull request merged**